sw/device/silicon_creator/rom/e2e/rom_e2e_c_init_test.c

```diff
@@ -0,0 +1,184 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/csr.h"
+#include "sw/device/lib/base/hardened.h"
+#include "sw/device/silicon_creator/lib/drivers/otp.h"
+#include "sw/device/silicon_creator/lib/manifest_def.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "otp_ctrl_regs.h"
+#include "pinmux_regs.h"
+#include "uart_regs.h"
+
+enum {
+  /**
+   * Base address of the pinmux registers.
+   */
+  kPinmuxBase = TOP_EARLGREY_PINMUX_AON_BASE_ADDR,
+  /**
+   * Base address of the UART registers.
+   */
+  kUartBase = TOP_EARLGREY_UART0_BASE_ADDR,
+};
+
+/**
+ * Triggers an instruction access fault.
+ */
+static void fault(void) {
+  ((void (*)(void))TOP_EARLGREY_SRAM_CTRL_MAIN_RAM_BASE_ADDR)();
+}
+
+/**
+ * Triggers an instruction access fault if the condition is not true.
+ */
+#define CHECK(condition) \
+  do {                   \
+    if (!(condition)) {  \
+      fault();           \
+    }                    \
+  } while (false)
+
+/**
+ * Check that the pinmux has been initialized correctly.
+ *
+ * Expected configuration:
+ * - If the OTP BOOTSTRAP_EN option is set, then pinmux should be configured
+ *   with the following peripheral/pad mapping, with pull-down enabled for all
+ *   pads:
+ *     * GPIO22 -> IOC0
+ *     * GPIO23 -> IOC1
+ *     * GPIO24 -> IOC2
+ * - Regardless of the OTP values, the pinmux should be configured with the
+ *   following mapping for UART0:
+ *     * UART RX -> IOC3
+ *     * IOC4 -> UART TX (output)
+ */
+static void pinmux_init_test(void) {
+  uint32_t bootstrap_en =
+      otp_read32(OTP_CTRL_PARAM_OWNER_SW_CFG_ROM_BOOTSTRAP_EN_OFFSET);
+  if (bootstrap_en == kHardenedBoolTrue) {
+    // GPIO 22 (input 22) -> IOC0 (MIO pad 22), pad attributes: pull-down
+    // enabled
+    uint32_t insel_gpio22 =
+        abs_mmio_read32(kPinmuxBase + PINMUX_MIO_PERIPH_INSEL_22_REG_OFFSET);
+    CHECK(insel_gpio22 == kTopEarlgreyPinmuxInselIoc0);
+    uint32_t attr22 =
+        abs_mmio_read32(kPinmuxBase + PINMUX_MIO_PAD_ATTR_22_REG_OFFSET);
+    // TODO: why is this attribute not legal on FPGA?
+    // CHECK(bitfield_bit32_read(attr22, PINMUX_MIO_PAD_ATTR_22_PULL_EN_22_BIT)
+    // == true);
```

Comment:
@jadephilipoom I'm not really sure, maybe it's not support the on the FPGA? @tjaychen @msfschaffner could you please clarify?

If that's the case, we can gate this check with
```
if (kDeviceType != kDeviceFpgaCw310) {
...
}
```
or something similar.

Reply:
@tjaychen @msfschaffner ping

Reply:
Yeah the reason why these pad attributes have WARL behavior is that the supported set varies quite a bit between ASIC, opensource and FPGA.

If you are curious what is supported, you can look at `prim_generic_pad_attr.sv` and `prim_xilinx_pad_attr.sv`:
https://github.com/lowRISC/opentitan/blob/10e7b2251fc01f0bbf2934804f466310810c2f2f/hw/ip/prim_generic/rtl/prim_generic_pad_attr.sv#L16-L67
https://github.com/lowRISC/opentitan/blob/10e7b2251fc01f0bbf2934804f466310810c2f2f/hw/ip/prim_xilinx/rtl/prim_xilinx_pad_attr.sv#L17-L51

As you can see, the FPGA (Xilinx) version only supports virtual open drain and inversion, but no dynamic pull ups / downs. Hence, gating this check based on target platform makes sense to me.

Reply:
(and apologies for the delay, this fell through the cracks).

Reply:
Thanks for confirming @msfschaffner!